### ISSUE TITLE
Caching of parsed OpenAPI definitions

### ISSUE THREAD

**Issue by judgej** [lezhnev74/openapi-psr7-validator#11]
This is more a question than anything. Hopefully it is covered somewhere.

So, for every inbound request to an API, the payload needs to be validated against an OpenAPI definition. That definition needs to be parsed/lexed/whatever into some form that can be used to do the validation. That will take time and resources to do.

So, is that parsed form of the OpenAPI definition cached? Can it even be cached? Does the full parsing happen on every API call, maybe it's not as resource intensive as I think it may be (but judging by the speed of client-based OpenAPI editors once a 3.0 definition goes above a few hundred lines, I would assume it is fairly intense).

I haven't tried it yet, but it looks ideal for a project I'm working on now.

**Comment by lezhnev74** [comment]
Good point! Indeed, performance is important. So this package is based on wonderful https://github.com/cebe/php-openapi by @cebe which does "reading" part. I think it should cache the schema between reads. 

I am going to dig deeper into this.

**Comment by judgej** [comment]
I was wondering if underlying packages did their own caching, which it would make sense if they did. But I would guess a caching object or factory would still need to be supplied. Thanks for the quick response - looking forward to trying it out.

**Comment by lezhnev74** [comment]
Quick update:
- cebe's `cebe/php-openapi` does not have a caching layer (I did not see one)
- `symfony/yaml`'s parser is used for actually reading YAML files. It does not have a caching layer as well.

Given that I am correct about the above, I think I should add a caching layer into the package.

**Comment by judgej** [comment]
Just out of curiosity, what would actually be cached? For example, PHP scripts are parsed into opcodes at runtime, and the opcodes are cached, so the parsing stage can be skipped. Do these libraries do something similar? Maybe they don't, and there is nothing that can be cached. I can't imagine a hash of the API requests having a particularly high hit-rate (assuming the cached result is a go/no-go result).

If there is an intermediate form for the OpenAPI definition - perhaps a data structure or object that can be serialised - then [PSR-6](https://www.php-fig.org/psr/psr-6/) support would be awesome.

**Comment by lezhnev74** [comment]
You read my mind about psr-6. 
Indeed there is an intermediary form. Text is converted to an object `\cebe\openapi\spec\OpenApi` which contains all parsed values. I believe this is a perfect candidate to be cached. Thus we save CPU cycles on parsing and converting openapi to that representation.

My plan is to serialize the object and store into the cache storage and then read and deserialize it back. I wonder if the mentioned process is better in terms of performance...

**Comment by lezhnev74** [comment]
I've done one iteration over this caching thing. See readme: https://github.com/lezhnev74/openapi-psr7-validator#caching-layer--psr-6-support

**Comment by judgej** [comment]
I'm going to use that awesome word again: *awesome*. I'm putting together a tiny API on stoplight.io to give this a first run through. Thank you @philsturgeon for pointing us here.

**Comment by lezhnev74** [comment]
👍 Please feel free to post any issues/ideas you encounter with this package.

**Comment by scaytrase** [comment]
I've looked up into this implementation and it has flaws for me.

Currently it caches only ready-for-use schema source data, but there is no real feedback about whether it is cached or not, so if your data source is not raw YAML or file (in our case its some doctrine annotations and some php code generating it) - than we should generate schema in order to use cached cebe\Schema at the end, so generation is wasted cpu here

**Comment by scaytrase** [comment]
From my point of view half of current `Validator` class content is something, that should be `ValidatorBuilder` at the end, and another part are schema utility helpers, which could be used with composition, not inheritance.

So It could be used like this (namings are reference only, just for example)

```php
final class YamlFileSchemaFactory implements CachecableSchemaFactory {
    public function createSchema(): /cebe/Schema { /* */ }

    public function getCacheKeyForSchema(): stirng { return crc32(...) }
}

interface CachecableSchemaFactory extends SchemaFactory {
    public function getCacheKeyForSchema(): string;
}

interface SchemaFactory {
    public function createSchema(): /cebe/Schema;
}

$schemaFactory = new YamlFileSchemaFactory('/path/to/yaml');

$validator = (new ValidatorBuilder())
    ->setCache($cacheItemPool) // optional
    ->setCacheKey($key) // optional if schema factory is cacheable
    ->setSchemaFactory($factory) // optional, see below
    ->fromYaml($yaml) // sets factory to built-in yaml string factory
    ->fromYamlFile($yamlFile) // sets factory to built-in yaml file factory
    ->fromJson($json) // same here
    ->fromJsonFile($jsonFile) // same here
    ->getServerRequestValidator() // or
    ->getResponseValidator()
```

builder internally checks given cache with given key (or generates it for cacheable factories) and if schema is not cached - then calls schema factory to obtain it. Predefined helpers could be used for bootstrapped usage, like

```php
$validator = (new ValidatorBuilder())
    ->setCache($cacheItemPool)
    ->fromYamlFile($yamlFile)
    ->getServerRequestValidator() 
```

I think it's easy to understand and control

**Comment by scaytrase** [comment]
I can try to provide PR with implementation to discuss a bit later

**Comment by lezhnev74** [comment]
> could be used with composition, not inheritance

I love this. Please do a PR, let's discuss later.

Indeed there are different concerns combined in the single Validator object at the moment - caching, building(reading openapi), actual validation. It works as a first iteration on the package but can be improved in many ways.

**Comment by scaytrase** [comment]
implemented initial thoughts as #16

**Comment by cebe** [comment]
> Quick update:
> 
> * cebe's `cebe/php-openapi` does not have a caching layer (I did not see one)
> * `symfony/yaml`'s parser is used for actually reading YAML files. It does not have a caching layer as well.
> 
> Given that I am correct about the above, I think I should add a caching layer into the package.

I can confirm there is no caching built into the lower layer packages, which is mainly because the library has no information on how the parsed/read content is going to be used, so there is no information on when to invalidate the cache.

There is [one issue with accessing an external URL multiple times while resolving references](https://github.com/cebe/php-openapi/issues/13), which I am going to solve inside of my lib, but all other caching should be done on the higher level, i.e. by openapi-psr7-validator.
